httomo/runner/dataset_store_interfaces.py

```diff
@@ -48,68 +50,131 @@ class DataSetSource(Protocol):
         """Slicing dimension - 0, 1, or 2"""
-        ...
+        ...  # pragma: no cover
+        
+    @property
+    def darks(self) -> np.ndarray:
+        """Darks array"""
+        ...  # pragma: no cover
+        
+    @property
+    def flats(self) -> np.ndarray:
+        """Flats array"""
+        ...  # pragma: no cover
 
     def read_block(self, start: int, length: int) -> DataSetBlock:
         """Reads a block from the dataset, starting at `start` of length `length`,
         in the current slicing dimension. Note that `start` is chunk-based,
         i.e. mean different things in different processes."""
-        ...
+        ...  # pragma: no cover
 
     def finalize(self):
         """Method intended to be called after reading all blocks is done,
         to give implementations a chance to close files, free memory, etc."""
-        ...
+        ...  # pragma: no cover
 
 
 class DataSetSink(Protocol):
     @property
     def global_shape(self) -> Tuple[int, int, int]:
         """Global data shape across all processes that we eventually have to write."""
-        ...
+        ...  # pragma: no cover
 
     @property
     def chunk_shape(self) -> Tuple[int, int, int]:
         """Returns the shape of a chunk, i.e. the data processed in the current
         MPI process (whether it fits in memory or not)"""
-        ...
+        ...  # pragma: no cover
 
     @property
     def chunk_index(self) -> Tuple[int, int, int]:
         """Returns the start index of the chunk within the global data array"""
-        ...
+        ...  # pragma: no cover
 
     @property
     def slicing_dim(self) -> Literal[0, 1, 2]:
         """Slicing dimension - 0, 1, or 2"""
-        ...
+        ...  # pragma: no cover
 
     def write_block(self, dataset: DataSetBlock):
         """Writes a block to the store, starting at the index in dataset.chunk_index,
-        in the current slicing dimension."""
-        ...
+        in the current slicing dimension.
+
+        NOTE: Implementers should make sure to move the dataset to CPU if required -
+        it may be on GPU when this method is called."""
+        ...  # pragma: no cover
 
     def finalize(self):
         """Method intended to be called after writing all blocks is done,
         to give implementations a chance to write everything to disk and close the file,
         etc."""
-        ...
+        ...  # pragma: no cover
+        
+class StoreBasedDataSetSink(DataSetSink):
+    """Interface for a DataSetSink that is store-based, i.e. where it's possible to construct
+       a reader from the same data store that this sink has been written first"""
+       
+    @abc.abstractmethod
+    def make_reader(
+        self, new_slicing_dim: Optional[Literal[0, 1, 2]] = None
+    ) -> DataSetSource:
+        """Method to make a source from this sink, which will read the data that was written,
+        possibly in a new slicing dimension"""
+        ...  # pragma: no cover
```

Comment:
Apologies for being dumb, I think we did discuss stuff in this area, but I'm still trying to understand things when reading through. I'm sort of struggling to understand the utility of the `StoreBasedDataSetSink` interface. 

`StoreBasedDataSetSink` adds a `make_reader()` method on top of `DataSetSink`, but `DataSetStoreWriter` already had a `make_reader()` method in its implementation of the `DataSetSink` interface. So I'm sturggling to see the difference between:
- `DataSetStoreReader` that already existed
- `StoreBasedDataSetSink` added in this change

as far as their usage in the task runner is concerned.

Also, given its name and docstring, `StoreBasedDataSetSink` seems to be somewhat tied to the data store, and the data store is a _particular implementation_ of the source and sink interfaces. In other words, `StoreBasedDataSetSink` is an interface (an abstraction), but is somewhat connected to or dependent on a specific implementation of other interfaces (a concretion), which seems a bit odd?

Any comments to help me reconcile my confusion would be great, and sorry again for being slow to grasp things!

Reply:
The reason for this is in [task_runner.py](https://github.com/DiamondLightSource/httomo/pull/205/files/726c9ab12ec9eccab089cc59265a12b5673daa91#diff-d0196284273478c424c5d41d96d3d67518f3a6e9cda84e282da786d380407fe9R45) - it needs to call `make_reader` for processing the next section, and technically it doesn't need to know if that's a store based writer or whatever it is. All it needs to know is that it's a sink interface with a `make_reader` method.

However, note that this is still work in progress - you will see that  `TaskRunner` actually instantiates `DataSetStoreWriter` explicitly, so it knows about it anyway. So your point is valid if it stays that way - at the moment it's not actually hiding implementations for `TaskRunner`. But that might change before the PR is ready to merge - still thinking about how.

Reply:
Ok, this makes sense that, due to being a work-in-progress, there's still a concrete dependence of `DataSetStoreWriter` in `TaskRunner`.

I guess that an interface which adds a `make_reader()` method on top of `DataSetSink` would still be needed in the future when the task runner is able to work with _any_ writer as long as a reader can be created from the writer. 

Currently, the interface that outline this "writer + can create reader from" behaviour is called `StoreBasedDataSetSink`. Which is a fine name for now, because the one and only implementation of that interface (`DataSetStoreWriter`) is indeed a sink that's based on a store.

However, given that in the future, a sink implementing a writer interface that requires a `make_reader()` method does not need to be stored-based necessarily, I think it'd be worth changing the name `StoreBasedDataSetSink` to something that doesn't refer to "store".

Reply:
Do you have a suggestion for how to name things?

Reply:
It's a very good question... :sweat_smile:

The fundamental difference between `DataSetSink` and `StoreBasedDataSetSink` is that the latter describes a sink that also allows the creation of a reader from it, isn't it? Maybe something like `ReadableSink`?

I'm not sure though, it's a name that makes some sense to me based on what behaviour the interface describes, but that doesn't mean it'd be sufficiently explanatory to others.

If we can't come up with a better name for it, I'm content with leaving it as `StoreBasedDataSetSink` for now, and to perhaps rename it in the future, possibly when we go over the data object hierarchy and rename stuff like `FullFileDataSet`.

Reply:
done